### ISSUE TITLE
[BUDI-7440] Creating dynamic form components

### ISSUE THREAD

**Issue by melohagan** [Budibase/budibase#5658]
### Discussed in [https://github.com/Budibase/budibase/discussions/5640](https://github.com/Budibase/budibase/discussions/5640)

<div type='discussions-op-text'>

<sup>Originally posted by **cirillojon** April 28, 2022</sup>
I am trying implement an order form where a user can input their own information while using the form and add new rows as they see fit without pulling data from a data provider. Currently we have a solution in a different workflow, but would like to implement it using budibase. Is there a functionality that can help us accomplish this?

![](https://uploads.linear.app/7ecfa297-cc77-4fdf-9117-fa5afdb8f908/181c9a73-849f-4a4b-ba74-957173e32834/962f8517-8ca2-4e39-b0cc-9b4f5e01997c?signature=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJwYXRoIjoiLzdlY2ZhMjk3LWNjNzctNGZkZi05MTE3LWZhNWFmZGI4ZjkwOC8xODFjOWE3My04NDlmLTRhNGItYmE3NC05NTcxNzNlMzI4MzQvOTYyZjg1MTctOGNhMi00ZTM5LWIwY2MtOWI0ZjVlMDE5OTdjIiwiaWF0IjoxNjkzNDczOTgwLCJleHAiOjE2OTM1NjAzODB9.di4zoqxhnFgm4uaJCy0oDxObq6kOB5YSBCRsV5lz9lU)

Another related discussion here: [https://github.com/Budibase/budibase/discussions/5189](https://github.com/Budibase/budibase/discussions/5189)

Would being able to use App State as a source for a Data Provider allow you to accomplish this? I.e. update state with a JSON object that represents your form elements, and then use a repeater to add them.

<sub>From [SyncLinear.com](https://synclinear.com) | [BUDI-7440](https://linear.app/budibase/issue/BUDI-7440/creating-dynamic-form-components)</sub>

**Comment by melohagan** [comment]
Just to add to this, you would also need to allow form field names to be bindable for this to work. Alternatively you could auto append a number on the end of the name for each input added.

![Screenshot 2022-05-24 at 13 49 24](https://user-images.githubusercontent.com/101575380/170038519-adcd5a50-9b21-4290-9667-c141817e6322.png)

**Comment by suparagroup** [comment]
@melohagan thank you for the report (was looking to report the same).
Did you happen to have reached a workaround that you can share?

**Comment by melohagan** [comment]
Hi @suparagroup This is a workaround I had tried using Form Steps: https://github.com/Budibase/budibase/discussions/5640#discussioncomment-2662603

It's a stack though - so you can only add an element onto the end, and remove the last element. You also need to know the maximum number of inputs.

**Comment by Shadowfita** [comment]
Making the form field names bindable would open up a lot of doors and make budibase a lot more powerful. +1 for that idea.

**Comment by TheFunkyBiskit** [comment]
Not sure if it's in bad taste to resurrect this thread, but this would be a super powerful feature.

**Comment by dannyshin76** [comment]
Actually there is a workaround using custom multi-picker as the data source. Although all subform elements need to be managed manually with onChange action (tedious as hell), everything is possible (unlimited subforms, adding, removing, re-ordering, etc.) I can share the how if anyone is interested.

**Comment by TheFunkyBiskit** [comment]
@dannyshin76 Please do!

**Comment by melohagan** [comment]
To add to this feature request further, it has come up a couple of times that users would like to be able to build Dynamic Forms from a schema, such as from a REST query:

https://github.com/Budibase/budibase/discussions/10121

**Comment by Simoon-F** [comment]
> Hi @suparagroup This is a workaround I had tried using Form Steps: [#5640 (comment)](https://github.com/Budibase/budibase/discussions/5640#discussioncomment-2662603)
> 
> It's a stack though - so you can only add an element onto the end, and remove the last element. You also need to know the maximum number of inputs.

How do I add a number after a field? I have N pieces of data that need to be rendered  drop-down selection box,
However, because the field names are the same, there is only one field when submitting the form. I tried to add
q{{ Duties Repeater.Duties.Auto ID}}
My ideal is = q1, but when I submit the form, I only have a value of q1.

![image](https://github.com/Budibase/budibase/assets/48503662/d8454697-eff0-43f8-b436-e850f1abb2a7)

**Comment by Simoon-F** [comment]
> Making the form field names bindable would open up a lot of doors and make budibase a lot more powerful. +1 for that idea.

This is a good function. 

Is there a plan? @melohagan

**Comment by Nezteb** [comment]
> users would like to be able to build Dynamic Forms from a JSON schema

@melohagan I listed a few open-source tools that might be useful for this functionality in https://github.com/Budibase/budibase/issues/10860 😄

**Comment by the-original-jellytree** [comment]
I agree, this is needed. I want to be able to have the user add an extra row if they need to input another line item (we have multiple line item entries per form).

**Comment by ChuTungLinh** [comment]
would the "bindable field name" be implemented in the future ? 
i have a survey page but i get the question dynamicly using rest API but i ran into the problem of submitting the answers since there's no way to differentiate which answer belong to which question, maybe i can work around it with JS array but it would 
 change every questions's answer on display to the newly selected answer which will probably confuse the users.

**Comment by nutgood** [comment]
+1 on this, it would be immensely useful.

**Comment by ConorWebb96** [comment]
Another discussion related to dynamic form components link [here](https://github.com/Budibase/budibase/discussions/11979).

**Comment by ConorWebb96** [comment]
Going to re-open as this needs to be discussed.

**Comment by melohagan** [comment]
Difficulty: Hard

Dynamic form components (where authors can define rows/entries on the fly without a single source table) will likely require treating App State as a data provider, so repeaters/forms can be backed by arbitrary JSON. We’d need to build an editor for defining those dynamic fields, sync them to state, and ensure the form can add/remove items while still supporting validation/binding. That touches state management, repeater binding logic, and the form controls, so it is a broader feature change.
